Modules/Module_Processing/Module_ASL/xASL_wrp_ResampleASL.m

```diff
@@ -254,84 +253,90 @@ function xASL_wrp_ResampleASL(x)
+            [ControlIm, LabelIm] = xASL_quant_GetControlLabelOrder(ASL_im);
+            PWI4D = ControlIm - LabelIm;
+            
+            % Skip every other value in x.Q.Initial_PLD as it was stored for both control and label images 
+            % we need the PLD vector now for the pairwise subtractions only            
+            initialPLD_PWI4D = initialPLD(1:2:end);
+            % Do the same for LabDur
+            LabDuration_PWI4D = LabDuration(1:2:end);
+
 			% After averaging across PLDs, we'll obtain these unique PLDs+LD combinations
-			% indexNew lists for each original position to where it should be averaged
-			[~, ~, indexNew] = unique([Initial_PLD_PWI, Initial_LabDur_PWI], 'stable', 'rows');
-			
-            % MultiPLD PWI after averaging
-			PWI = zeros(size(ASL_im,1), size(ASL_im,2), size(ASL_im,3), max(indexNew)); % preallocate PWI
-            for nPLD = 1:max(indexNew)
-                PWI(:, :, :, nPLD) = xASL_stat_MeanNan(ASL_im(:, :, :, indexNew == nPLD), 4); % Averaged PWI4D 
+			% indexAverage_PLD_LabDur lists for each original position to where it should be averaged
+			[~, ~, iUnique_PLD_LabDur] = unique(unique([initialPLD_PWI4D, LabDuration_PWI4D], 'stable', 'rows'), 'stable', 'rows');
+
+            % MultiPLD-multiLabDur PWI3D after averaging
+			PWI3D = zeros(size(PWI4D,1), size(PWI4D,2), size(PWI4D,3), max(iUnique_PLD_LabDur)); % preallocate PWI
+            for iPLD_LabDur = 1:max(iUnique_PLD_LabDur)
+                PWI3D(:, :, :, iPLD_LabDur) = xASL_stat_MeanNan(PWI4D(:, :, :, iUnique_PLD_LabDur == iPLD_LabDur), 4); % Averaged PWI4D 
             end
-            
-            % Save PWI
-            fprintf('%s\n', PathPWI4D{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI4D{iSpace}, PWI, 32, false);
-            
-            PWIsingle = xASL_stat_MeanNan(ASL_im, 4); % create single PWI for further steps in ASL module
-            
-            % Save single PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWIsingle, 32, false);
-            
         else
-            PWI = xASL_stat_MeanNan(ASL_im, 4); % singlePLD PWI
-            % Save PWI
-            fprintf('%s\n', PathPWI{iSpace});
-            xASL_io_SaveNifti(PathASL{iSpace}, PathPWI{iSpace}, PWI, 32, false);
-        end            
+            PWI4D = ASL_im;
+        end
+        
+        PWI3D = xASL_stat_MeanNan(PWI4D, 4);
     end
+    
+    
+    % =====================================================================
+    % D) Save subtracted to disk
+
+    % Save PWI (single volume)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI{iSpace}, PWI, 32, false);
+
+    % Save PWI3D (averaged volume for each PLD-labdur combination)
+    fprintf('%s\n', PathPWI{iSpace});
+    xASL_io_SaveNifti(PathASL4D{iSpace}, PathPWI3D{iSpace}, PWI3D, 32, false);
+    
+    % Save PWI4D (subtracted only, not yet averaged)

```

Comment:
Currently, we do the averaging, so we need to then make sure in the second step that all averaging is removed. 

This is especially valid for Time-encoded and multi-TE where this averaging is a bit more complex and isn't a pure average only.

Reply:
Yes, the averaging should go to the quantification module.

I need to check how this goes with the temporary PWI needed for registration, we might want to call part of this function separately within other functions that need this temporary PWI (e.g., in xASL_wrp_Realign & xASL_wrp_Registration)

Currently, this code is duplicated there for Hadamard for registration and motion correction.

Reply:
We don't average now. Can be resolved.